LFS-501: Cardiac Questionnaires import

nela-jankechova opened the pull request:
Import of questionnaires from [this spreadsheet in tsv format](https://docs.google.com/spreadsheets/d/1rFITfjJT9tkan2lWU4XUZTeKpeavW37SPhYc31lsugw/edit?usp=sharing ).

These subjectType of these questionnaires will need to be set.

[Jira link](https://phenotips.atlassian.net/browse/LFS-501).

nela-jankechova committed:
LFS-501: Add JSON files

nela-jankechova committed:
LFS-501: Add python script

veronikaslc commented on questionnaires/clinical-data/util/heartCentreData.py:
```diff
@@ -0,0 +1,225 @@
+        for row in reader:
+            question = row['nameShort'].strip()
+            if question:
+                questionnaire[question] = {}
```
probably don't need this line

veronikaslc commented on questionnaires/clinical-data/util/heartCentreData.py:
```diff
@@ -0,0 +1,225 @@
+
+# Converts the data type in 'UserFormatType' to one supported in LFS
+def switch_data_type(argument):
+    switcher = {
```
`switcher` better be defined as a global const

veronikaslc replied:
also function name might be better in a form of `convert_to_LFS_data_type` and `switcher` name better in a form of `data_to_LFS_types`

veronikaslc commented on questionnaires/clinical-data/util/heartCentreData.py:
```diff
@@ -0,0 +1,225 @@
+        'cat list': 'text',
+        'Text': 'text',
+        'Date': 'date',
+        'Integer': 'long'
```
Missed type `Age (Months:Days)`
I'd also add `'Yes,No': 'boolean'`

veronikaslc commented on questionnaires/clinical-data/util/heartCentreData.py:
```diff
@@ -0,0 +1,225 @@
+                questionnaire[question] = {
+                    'jcr:primaryType': 'lfs:Question',
+                    'text': row['nameFull'].strip() or question,
+                    'dataType': switch_data_type(row['UserFormatType'])
```
I'd pass on the `row['Categorical list']` here instead of later in a form of 
`switch_data_type(row['UserFormatType'] or row['Categorical list'])`
since both arguments could decide the value of `dataType`. No need to apply `options_list()` since full strings will be compared.

veronikaslc commented on questionnaires/clinical-data/util/heartCentreData.py:
```diff
@@ -0,0 +1,225 @@
+        'Date': 'date',
+        'Integer': 'long'
+    }
+    return switcher.get(argument, 'text')
```
Since you have a default, you can miss all `text` options in `switcher` const

veronikaslc replied:
I'd add `.strip()` to the `argument` and compare all in lower case just to be safe.

veronikaslc commented on questionnaires/clinical-data/util/heartCentreData.py:
```diff
@@ -0,0 +1,225 @@
+                    'text': row['nameFull'].strip() or question,
+                    'dataType': switch_data_type(row['UserFormatType'])
+                }
+                if row['RangeMinValid'] != '' and row['RangeMinValid'] != None:
```
Shorter more reliable version `if questionnaire['dataType'] == 'long'`

OrdiNeu replied:
Not every `long` question has a data range, unfortunately

veronikaslc commented on questionnaires/clinical-data/util/heartCentreData.py:
```diff
@@ -0,0 +1,225 @@
+                if row['MissingData']:
+                    insert_min_answers(questionnaire[question], row)
+                if row['Categorical list']:
+                    if len(options_list(row['Categorical list'])) == 1:
```
Not needed if `row['Categorical list']` is analyzed above

veronikaslc commented on questionnaires/clinical-data/util/heartCentreData.py:
```diff
@@ -0,0 +1,225 @@
+                    insert_range(questionnaire[question], row)
+                if row['MissingData']:
+                    insert_min_answers(questionnaire[question], row)
+                if row['Categorical list']:
```
`if row['Categorical list'] and not switcher.get(row['Categorical list']) and not row['Categorical list'].endswith('Other'):`
Lines 202 - 204 are not needed if `row['Categorical list']` is analyzed above

veronikaslc commented on questionnaires/clinical-data/util/heartCentreData.py:
```diff
@@ -0,0 +1,225 @@
+
+# Creates lfs:AnswerOptions from the CSV in 'Categorical List'
+def insert_options(question, row):
+    option_list = options_list(row['Categorical list'])
```
I didn't see any '(' or ')' in `Categorical list` column. Also it seems that the only used separator is `,`. So the whole function could be substituted with 
`option_list = [x.strip().replace('/', '-') for x in categorical_list.split(',')]`
Lines 130-143 are not needed.

acrowthe replied:
QIVariables has catagorical list entries in the format ` (yes/no)` meaning the simplification can't be used

veronikaslc commented on questionnaires/clinical-data/util/heartCentreData.py:
```diff
@@ -0,0 +1,225 @@
+                        questionnaire[question].update({'dataType': switch_data_type(row['Categorical list'])})
+                    else:
+                        insert_options(questionnaire[question], row)
+                if row['Description'] != '':
```
`if row['Description']:`
  `questionnaire[question].update({ 'description': row['Description'].strip() })`

veronikaslc commented on questionnaires/clinical-data/util/heartCentreData.py:
```diff
@@ -0,0 +1,225 @@
+                    insert_description(questionnaire[question], row)
+                if row['UserFormatType'] == 'Text (categorical list)' or row['UserFormatType'] == 'cat list':
+                    questionnaire[question].update({'displayMode': 'list'})
+                    print(row)
```
don't need this line

sashaandjic commented:
-  "Other" option is missing in all multi choice lists ( e.g. for O-5 Testing field Language it should provide Spanish and Other; only Spanish is available)
- complete lack of conditional sections and fields

OrdiNeu committed:
LFS-501: Cardiac Questionnaires import
Clean up heartCentreData.py

OrdiNeu committed:
LFS-501: Cardiac Questionnaires
Further cleanup

OrdiNeu committed:
LFS-501: Cardiac Questionnaires import
Rename QIVariables to Q1Variables, as per the sheet
Further code cleanup

OrdiNeu committed:
LFS-501: Cardiac Questionnaires
Fix conditional sections

OrdiNeu committed:
LFS-501: Cardiac Questionnaires Import
Fix the Other category

acrowthe committed:
LFS-501: Cardiac Questionnaires import
Add runMode cardiac to enable cardiac forms
- Adjust runMode noforms to support this feature

acrowthe committed:
LFS-501: Cardiac Questionnaires import
Move remaining lfs resources behind runMode forms
- SubjectTypes Tumor and Tumor region
- Variants extensions

sdumitriu commented:
@acrowthe Since LFS-492 has been merged first, can you run [the script](https://github.com/ccmbioinfo/lfs/blob/1c91f5ca1f6dd6bc36f7625541ef39aba24f5c84/Utilities/JSON-to-XML/json_to_xml.py) to convert all questionnaires into XML format?

acrowthe commented:
@sdumitriu Will do

acrowthe committed:
LFS-501: Cardiac Questionnaires import
Final adjustments before review
- Fix variants sidebar showing in noforms mode
- Convert questionnaires to xml using Utilities/JSON-to-XML
- Remove Q1Variables file as it was re-created as QIVariables

acrowthe committed:
LFS-501: Cardiac Questionnaires import
Address python review comments

acrowthe commented:
Runmode testing instructions:

LFS
`rm -rf ./sling/; java -jar distribution/target/lfs-*.jar`
- With no runmodes specified, only non-specific and LFS resources should show up
  - subject types Patients, Tumors, Tumor Region
  - All LFS Forms (chemotherapy, demographics, radiotherapy, tumor, somatic variants, tissue metrix), no cardiac forms
  - Variants entry in the sidebar

Cardiac
`rm -rf ./sling/; java -jar distribution/target/lfs-*.jar -Dsling.run.modes=noforms,cardiac`
- With the runmodes `noforms` and `cardiac specified, no only non-specific and cardiac resources should show up
  - subject type Patient
  - 3 variables forms: 0-5ND, 6-21ND, QI

acrowthe committed:
LFS-501: Cardiac Questionnaires import
Fix conditional questions not checking boolean answers correcctly

acrowthe committed:
LFS-501: Cardiac Questionnaires import
Make question names lowercase for better conditional import

acrowthe committed:
LFS-501: Cardiac Questionnaires import
Fix conditional statements

acrowthe commented:
Known issue:
- "<form> not completed" messages on the 6-21NDVariables form do not show up at the right times

Looking into a solution, but the remainder of the functionality can be tested (conditional appearance of other questions, spelling, runmodes)

acrowthe committed:
LFS-501: Cardiac Questionnaires import
Fix incorrect conditionals on 6-21NDVariables test not complete questions
- Improved handling of `<>` operator

acrowthe commented:
Fixed incorrect conditionals

sdumitriu committed:
LFS-501: Cardiac Questionnaires import
Split "forms" and "lfs" runmodes, to preserve the ability to have an instance without any forms at all

sdumitriu committed:
[cleanup] Fixed whitespace
